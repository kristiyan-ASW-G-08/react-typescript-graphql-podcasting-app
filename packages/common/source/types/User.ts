export default interface User {
  username: string;
  handle: string;
  email: string;
  profilePhoto: string;
  headerPhoto: string;
  date: string;
  followers: number;
  avatar: string;
  cover: string;
}
