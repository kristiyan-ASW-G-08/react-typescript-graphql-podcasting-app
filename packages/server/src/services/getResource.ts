import { UserInputError } from 'apollo-server';
import mongoose, { Model, Document } from 'mongoose';

interface FindQuery {
  name: string;
  value: number | string | mongoose.Types.ObjectId;
}
const getResource = async <T extends Document>(
  model: Model<T>,
  { value, name }: FindQuery,
  Error: any,
  select: string = '',
): Promise<T> => {
  const resource = await model.findOne({ [name]: value }).select(select);
  if (!resource) {
    throw Error;
  }
  return resource;
};

export default getResource;
