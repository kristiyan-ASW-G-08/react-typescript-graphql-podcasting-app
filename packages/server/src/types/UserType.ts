import { Document } from 'mongoose';

export default interface UserType extends Document {
  username: string;
  email: string;
  password: string;
}
