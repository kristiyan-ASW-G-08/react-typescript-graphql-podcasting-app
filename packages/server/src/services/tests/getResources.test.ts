import User from '@src/users/UserModel';
import UserType from '@customTypes/UserType';
import getResource from '@services/getResource';

describe('getResources', () => {
  expect.assertions(4);
  afterEach(() => jest.clearAllMocks());
  afterAll(() => jest.restoreAllMocks());

  it('should call find and select', async () => {
    const select = jest.fn(() => true);
    const findQuery = { name: 'email', value: 'testMail@test' };
    const selectQuery = 'username email';
    const findOneSpy = jest
      .spyOn(User, 'findOne')
      // @ts-ignore
      .mockReturnValue({ select });
    await getResource<UserType>(User, findQuery, selectQuery);

    expect(findOneSpy).toHaveBeenCalledTimes(1);
    expect(findOneSpy).toHaveBeenCalledWith({
      [findQuery.name]: findQuery.value,
    });
    expect(select).toHaveBeenCalledTimes(1);
    expect(select).toHaveBeenCalledWith(selectQuery);
  });
  it('should throw an error if find and select resolve to falsy value', async () => {
    expect.assertions(1);
    const select = jest.fn(() => false);
    const findQuery = { name: 'email', value: 'testMail@test' };
    const selectQuery = 'username email';
    const findOneSpy = jest
      .spyOn(User, 'findOne')
      // @ts-ignore
      .mockReturnValue({ select });

    await expect(
      getResource<UserType>(User, findQuery, selectQuery),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
