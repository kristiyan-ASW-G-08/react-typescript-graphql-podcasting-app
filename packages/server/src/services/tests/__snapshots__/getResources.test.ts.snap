// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getResources should throw an error if find and select resolve to falsy value 1`] = `undefined`;
